DeviceCopy should have the same layout as input

yushangdi opened the pull request:
Summary: Fix https://github.com/pytorch/pytorch/issues/159612

- Fix the meta implementation of `nan_to_num`, it should preserve the stride of the input
- The DeviceCopy IR node should always preserve the input's layout, so we don't end up with a contiguous call during device copy

Test Plan:
```
buck2 run @mode/dev-nosan fbcode//caffe2/test/inductor:test_aot_inductor -- -r test_d2h_copy
```


Rollback Plan:

Differential Revision: D79411407


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

eellison reviewed (commented):
Inside of DeviceCopy, you can `ExternKernel.require_contiguous` the input to constrain its stride. And initialize the output with Fixed contiguous layout.

eellison commented on torch/_inductor/ir.py:
```diff
@@ -7005,11 +7005,15 @@ def create(cls, x: IRNode, device: torch.device, non_blocking: bool) -> IRNode:
+                x.get_dtype(),
+                x.get_size(),
+                x.get_stride(),
+                x.get_layout().offset,
```
The offset is not required to be contiguous and avoid copy.

eellison commented on torch/_inductor/ir.py:
```diff
@@ -7005,11 +7005,15 @@ def create(cls, x: IRNode, device: torch.device, non_blocking: bool) -> IRNode:
 
         developer_warning("DeviceCopy in input program")
         constant_args = (non_blocking,)
+        # Device Copy should keep the same layout as input
```
> Device Copy should keep the same layout as input

nit - could you add something that the aten cuda copy adds an intermediary if both inputs are not contiguous ? (even though it should just require them to have the same memory layout..)

eellison replied:
I think it is just contiguous.. but let know if you see otherwise somewhere

eellison reviewed (commented):
one comment on the offset but looks good otherwise.

yushangdi commented on torch/_meta_registrations.py:
```diff
@@ -7025,8 +7025,7 @@ def _amp_foreach_non_finite_check_and_unscale_(self, found_inf, inv_scale):
 def nan_to_num(self, nan=None, posinf=None, neginf=None):
-    result_size = list(self.size())
-    return self.new_empty(result_size)
+    return torch.empty_like(self)
```
or maybe we should fix `new_empty` to have the same stride as input? But I don't know if people're already relying on the existing behavior of `new_empty`

yushangdi committed:
DeviceCopy should have the same layout as input (#159615)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/159615

Fix https://github.com/pytorch/pytorch/issues/159612

Test Plan:
```
buck2 run mode/dev-nosan fbcode//caffe2/test/inductor:test_aot_inductor -- -r test_d2h_copy

python test/inductor/test_cudagraph_trees.py -k test_graph_partition_cpu_scalar_mutation

```

Reviewed By: desertfire

Differential Revision: D79411407

izaitsevfb commented:
@pytorchbot merge
